### PR TITLE
Signature + VRF spec

### PR DESCRIPTION
**Pull request by nikkolasg**
Main changes:
 - renamed secp256k1 to ECDSA (over the secp256k1 curve)
 - moved VRF inside algorithms/crypto since it fits just next to BLS (and is not a "Filecoin" scheme)
 - moved into separate file Signature, ECDSA , BLS and VRF

TODO: 
 - put in place the links between the .id files and .md files at the right place.

### CODE REVIEW

File: src/algorithms/crypto/vrf.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+Any third party in possession of the corresponding public key, the message and
+the VRF output can verify if the digest has been computed correctly and from the
+correct signer. Using a VRF in the ticket generation process allows anyone to
+verify if a block comes from an eligible block producer (see {{<sref tickets
+"Ticket Generation" >}} for more details).
+
+BLS signature can be used as the basis to construct a VRF. Filecoin transforms
+the BLS signature scheme it uses (see {{<sref signatures Signatures>}} into a
+VRF, Filecoin uses the random oracle model and deterministically hash the
+signature to produce the final digest. Filecoin uses SHA256 as the hash
```
Comment (sternhenri) [inline src/algorithms/crypto/vrf.md]: Q here of whether to simply use blake2s as you point out @nikkolasg. Change?

**Comment by jbenet** [comment]
@sternhenri @nikkolasg see https://github.com/filecoin-project/specs/pull/537/commits/f3361a394e3570d1913044be0a673568abdbe819 (`struct` use)

File: src/algorithms/crypto/signatures.md [hunk 2]
```diff
@@ -93,8 +68,53 @@ SignatureBytes = [0x30][len][0x02][r][indicator][s][indicator][recovery]
...
+- Prove knowledge of the secret key
+- Use a modified scheme (such as [BLS Multi Sig](https://crypto.stanford.edu/~dabo/pubs/papers/BLSmultisig.html))
+
+Fortunately, Filecoin can enforce the second condition to safely use the
+aggregation property:
+Filecoin uses aggregation only for aggregating the transaction's signature of a
+block. since Filecoin uses the account model to represent the state of the
+chain, each message for a given signer is used in combination of a nonce to
+avoid replay attacks. As a direct consequence, every transaction's message is
+unique thereby the aggregation is done on distinct messages.
```
[the remark this answers is not in the document]
Comment (jbenet) [reply src/algorithms/crypto/signatures.md]: yeah! great to have! 

@nikkolasg could we get the aggregate sig functions + verification defined in code too? that will make it super clear how to use them.

[the remark this answers is not in the document]
Comment (jbenet) [reply src/algorithms/crypto/signatures.md]: in particular, one Q that came up, can we verify a partial aggregate, or do we need all the messages? specifically, is the following possible?

```
sigs := []BLSSignature{s1, s2, s3, s4, s5, ..., s4000}
aggSig := BLSAggregateSignatures(sigs)

// then, later:
allPairs := []SigPair{ SigPair(pk1, m1), SigPair(pk2, m2), ... SigPair(pk4000, m4000)}
verifyAll := BLSAggregateVerifySignature(aggSig, allPairs)
if verifyAll { ... }

// but can we do:
somePairs := []SigPair{ SigPair(pk1, m1), SigPair(pk2, m2), ... SigPair(pk2000, m2000)}
verifySome := BLSAggregateVerifySignature(aggSig, somePairs)
if verifySome { ... }

somePairs2 := []SigPair{ SigPair(pk2001, m2001), SigPair(pk2002, m2002), ... SigPair(pk4000, m4000)}
verifyRest := BLSAggregateVerifySignature(aggSig, somePairs2)
if verifyRest { ... }
```
The above is relevant to do blackbox parallelizing of verifications

**Comment by jbenet** [comment]
LGTM -- left some comments, but im fine to merge as is and update later. as you prefer @nikkolasg

File: src/algorithms/crypto/vrf.md (hunk 1 above)
Comment (nikkolasg) [reply src/algorithms/crypto/vrf.md]: I'm pushing for blake2s yes, will change that

File: src/algorithms/crypto/signatures.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (nikkolasg) [reply src/algorithms/crypto/signatures.md]: Yes we can do that - I will put the code for for aggregating sig + public key and will make it clear we can have partial aggregation.

[the remark this answers is not in the document]
Comment (nikkolasg) [reply src/algorithms/crypto/signatures.md]: Although I can't compile yet so that'll be pseudo code, one of you mac users will have to make it work ;)

**Comment by sternhenri** [comment]
Let me know when you've added your changes @nikkolasg I can make sure it passes tests and merge it if of use! excited to get this in.

**Comment by nikkolasg** [comment]
I've added a small text about use of domain separation tags.
- We should precise a domain separation tag for the message of the BLS signature used in the ticket chain. I wanted to only talk about the VRF and BLS separately here so I did not put the tag for the BLS sig. This should be at the place where we describe the ticket chain: the message passes from `last_vrf_output` to `DST || last_vrf_output` where `DST` can be anything ("chain", or "bls" or whatever).
- In general this is something we should do as much as possible. I understand that it might not be possible always when using seck for compatibility reasons but any usage of a signature scheme with a given key should use a domain separation tag. It doesn't cost and it prevents certain class of attacks.

**Comment by zixuanzh** [comment]
there are two filcrypto, one in algorithms and one in libraries. Made the minimal fix to get this PR compiled to unblock others. Might want to take a look to make sure the naming is right cc @porcuquine.

**Pull request merged**